patches/2025-03-10-create-primary-name-request-demand-factor.lua

```diff
@@ -0,0 +1,238 @@
+	local existingRequest = primaryNames.getPrimaryNameRequest(initiator)
+	assert(
+		not existingRequest or existingRequest.name ~= name,
+		"Primary name request by '" .. initiator .. "' for '" .. name .. "' already exists"
+	)
+
+	--- check the primary name is not already owned
+	local primaryNameOwner = primaryNames.getAddressForPrimaryName(name)
+	assert(not primaryNameOwner, "Primary name is already owned")
+
+	local record = arns.getRecord(baseName)
+	assert(record, "ArNS record '" .. baseName .. "' does not exist")
+	assert(arns.recordIsActive(record, timestamp), "ArNS record '" .. baseName .. "' is not active")
+
+	local requestCost = arns.getTokenCost({
+		intent = "Primary-Name-Request",
+		name = name,
+		currentTimestamp = timestamp,
+		record = record,
+	})
+
+	local fundingPlan = gar.getFundingPlan(initiator, requestCost.tokenCost, fundFrom)
+	assert(fundingPlan and fundingPlan.shortfall == 0, "Insufficient balances")
+	local fundingResult = gar.applyFundingPlan(fundingPlan, msgId, timestamp)
+	assert(fundingResult.totalFunded == requestCost.tokenCost, "Funding plan application failed")
+
+	--- transfer the primary name cost from the initiator to the protocol balance
+	balances.increaseBalance(ao.id, requestCost.tokenCost)
+	demand.tallyNamePurchase(requestCost.tokenCost)
+
+	local request = {
+		name = name,
+		startTimestamp = timestamp,
+		endTimestamp = timestamp + constants.PRIMARY_NAME_REQUEST_DURATION_MS,
+	}
+
+	--- if the initiator is base name owner, then just set the primary name and return
+	local newPrimaryName
+	if record.processId == initiator then
+		newPrimaryName = primaryNames.setPrimaryNameFromRequest(initiator, request, timestamp)
+	else
+		-- otherwise store the request for asynchronous approval
+		PrimaryNames.requests[initiator] = request
+		primaryNames.scheduleNextPrimaryNamesPruning(request.endTimestamp)
+	end
+
+	return {
+		request = request,
+		newPrimaryName = newPrimaryName,
+		baseNameOwner = record.processId,
+		fundingPlan = fundingPlan,
+		fundingResult = fundingResult,
+		demandFactor = demand.getDemandFactorInfo(),
+	}
+end
+
+-- Now update main.lua to use the new function and add the demand factor data
+local createPrimaryNameRequestHandlerIndex = utils.findInArray(Handlers.list, function(handler)
+	return handler.name == "requestPrimaryName"
+end)
+
+if not createPrimaryNameRequestHandlerIndex then
+	error("Failed to find requestPrimaryName handler")
+end
+
+local createPrimaryNameRequestHandler = Handlers.list[createPrimaryNameRequestHandlerIndex]
+if not createPrimaryNameRequestHandler then
+	error("Failed to find requestPrimaryName handler")
+end
+
+local function Send(msg, response)
+	if msg.reply then
+		--- Reference: https://github.com/permaweb/aos/blob/main/blueprints/patch-legacy-reply.lua
+		msg.reply(response)
+	else
+		ao.send(response)
+	end
+end
+
+local function assertValidFundFrom(fundFrom)
```

Comment:
Don't all these local functions need to be hoisted up above their usage above? 🤔

Reply:
they are all above 

```lua
createPrimaryNameRequestHandler.handler = function(msg)
   // new handle function
```